# FIXED COUPON BOND - QL EXAMPLE

This is a based on example in http://gouthamanbalaraman.com/blog/quantlib-bond-modeling.html

In [2]:
import numpy as np
import matplotlib.pyplot as plt

In [3]:
from financepy.finutils import *
from financepy.products.bonds.Bond import *
from financepy.market.curves.FinDiscountCurveZeros import *

###################################################################
# FINANCEPY BETA Version 0.180 - This build: 22 Sep 2020 at 19:38 #
#     This software is distributed FREE & WITHOUT ANY WARRANTY    #
# For info and disclaimer - https://github.com/domokane/FinancePy #
###################################################################



# Define the Bond

In [8]:
issue_date = FinDate(15,1,2010)
maturity_date = FinDate(15, 1, 2016)
coupon = 0.06
freq_type = FrequencyTypes.SEMI_ANNUAL
accrual_type = DayCountTypes.THIRTY_360_BOND
face = ONE_MILLION

In [9]:
bond = Bond(issue_date, maturity_date, coupon, freq_type, accrual_type, face)

In [10]:
print(bond)

OBJECT TYPE: Bond
ISSUE DATE: FRI 15 JAN 2010
MATURITY DATE: FRI 15 JAN 2016
COUPON: 0.06
FREQUENCY: FrequencyTypes.SEMI_ANNUAL
ACCRUAL TYPE: DayCountTypes.THIRTY_360_BOND
FACE AMOUNT: 1000000


To see the cash flows we first need to set the settlement date of the bond. 

In [11]:
settlement_date = FinDate(15, 1, 2015)

In [12]:
bond.printFlows(settlement_date)

THU 15 JAN 2015      30000.00 
WED 15 JUL 2015      30000.00 
FRI 15 JAN 2016    1030000.00 


## Discounting Bond Flows

We wish to define a zero rate curve. For this we need the dates and values of the zero rates.

In [13]:
zeroDates = [FinDate(15,7,2015), FinDate(15,1,2016)]
zeroRates = [0.005, 0.007]

In [9]:
discount_curve = FinDiscountCurveZeros(settlement_date, zeroDates, zeroRates, 
                                      FrequencyTypes.ANNUAL,
                                      DayCountTypes.THIRTY_360_BOND)

In [10]:
print(discount_curve)

OBJECT TYPE: FinDiscountCurveZeros
VALUATION DATE: THU 15 JAN 2015
FREQUENCY TYPE: FrequencyTypes.ANNUAL
DAY COUNT TYPE: DayCountTypes.THIRTY_360_BOND
INTERP TYPE: FinInterpTypes.FLAT_FWD_RATES
DATES: ZERO RATES
WED 15 JUL 2015:  0.0050000
FRI 15 JAN 2016:  0.0070000



In [11]:
bond.clean_price_from_discount_curve(settlement_date, discount_curve)

105.27653992490681

In [12]:
bond.full_price_from_discount_curve(settlement_date, discount_curve)

105.27653992490681

As we are on the issue date of the bond there is no accrued.

This agrees with QL which finds a price of 105.27654

Copyright (c) 2020 Dominic O'Kane